redisvl/utils/rerank/voyageai.py

```diff
@@ -0,0 +1,195 @@
+        api_config: Optional[Dict] = None,
+    ) -> None:
+        """
+        Initialize the VoyageAIReranker with specified model, ranking criteria,
+        and API configuration.
+
+        Parameters:
+            model (str): The identifier for the VoyageAI model used for reranking.
+            rank_by (Optional[List[str]]): Optional list of keys specifying the
+                attributes in the documents that should be considered for
+                ranking. None means ranking will rely on the model's default
+                behavior.
+            limit (int): The maximum number of results to return after
+                reranking. Must be a positive integer.
+            return_score (bool): Whether to return scores alongside the
+                reranked results.
+            api_config (Optional[Dict], optional): Dictionary containing the API key.
+                Defaults to None.
+
+        Raises:
+            ImportError: If the voyageai library is not installed.
+            ValueError: If the API key is not provided.
+        """
+        super().__init__(
+            model=model, rank_by=rank_by, limit=limit, return_score=return_score
+        )
+        self._initialize_clients(api_config)
+
+    def _initialize_clients(self, api_config: Optional[Dict]):
+        """
+        Setup the VoyageAI clients using the provided API key or an
+        environment variable.
+        """
+        # Dynamic import of the voyageai module
+        try:
+            from voyageai import AsyncClient, Client
+        except ImportError:
+            raise ImportError(
+                "VoyageAI reranker requires the voyageai library. \
+                    Please install with `pip install voyageai`"
+            )
+
+        # Fetch the API key from api_config or environment variable
+        api_key = (
+            api_config.get("api_key") if api_config else os.getenv("VOYAGE_API_KEY")
+        )
+        if not api_key:
+            raise ValueError(
+                "VoyageAI API key is required. "
+                "Provide it in api_config or set the VOYAGE_API_KEY environment variable."
+            )
+        self._client = Client(api_key=api_key)
+        self._aclient = AsyncClient(api_key=api_key)
+
+    def _preprocess(
+        self, query: str, docs: Union[List[Dict[str, Any]], List[str]], **kwargs
+    ):
+        """
+        Prepare and validate reranking config based on provided input and
+        optional overrides.
+        """
+        limit = kwargs.get("limit", self.limit)
+        return_score = kwargs.get("return_score", self.return_score)
+        truncation = kwargs.get("truncation")
+
+        if all(isinstance(doc, dict) and "content" in doc for doc in docs):
+            texts = [
+                str(doc["content"])
+                for doc in docs
+                if isinstance(doc, dict) and "content" in doc
+            ]
+        else:
+            texts = [str(doc) for doc in docs]
```

Comment:
@tylerhutcherson I checked the errors in tests and probably it would be better to change this line to raise a ValueError instead, something like this:
`            raise ValueError(
                "The docs should either be List[str] or List[Dict[str, Any]] - "
                "this case, the documents to rerank has to be in the 'content'"
            )
`
What do you think?

Reply:
Agreed - going to address now and try to get this wrapped up.